File: server/util/pojos/note.js
// Nota privada de un usuario. Vive únicamente en MongoDB (documento 'notes'
// de la colección del usuario); nunca se envía al triple-store LOD.
//
// El dibujo se guarda como fichero en disco (drawingFile) y en el documento
// solo queda su nombre y su tamaño en bytes, para no agotar el límite de 16 MB
// por documento de MongoDB y poder calcular la cuota sin tocar el disco.
// El campo drawing (PNG en base64) se mantiene únicamente para poder seguir
// leyendo las notas creadas antes de este cambio.
class Note {
    constructor(data) {
        if (data === null || typeof data !== 'object') {
            throw new Error('No valid data for create Note');
        }

        if (data.id !== undefined && typeof data.id === 'string' && data.id.trim() !== '') {
            this._id = data.id.trim();
        } else {
            throw new Error('Note id');
        }

        this._title = typeof data.title === 'string' && data.title.trim() !== '' ? data.title.trim() : undefined;
        this._text = typeof data.text === 'string' && data.text.trim() !== '' ? data.text.trim() : undefined;
        // Dibujo guardado como fichero
        this._drawingFile = typeof data.drawingFile === 'string' && data.drawingFile.trim() !== '' ? data.drawingFile.trim() : undefined;
        this._drawingSize = typeof data.drawingSize === 'number' && data.drawingSize > 0 ? data.drawingSize : undefined;
        // Dibujo heredado: PNG en base64 dentro del propio documento
        this._drawing = typeof data.drawing === 'string' && data.drawing !== '' ? data.drawing : undefined;
        // Vínculo opcional con un lugar
        this._idPlace = typeof data.idPlace === 'string' && data.idPlace.trim() !== '' ? data.idPlace.trim() : undefined;
        this._labelPlace = typeof data.labelPlace === 'string' && data.labelPlace.trim() !== '' ? data.labelPlace.trim() : undefined;

        this._creation = typeof data.creation === 'number' ? data.creation : Date.now();
        this._lastUpdate = typeof data.lastUpdate === 'number' ? data.lastUpdate : this._creation;
    }

    get id() { return this._id; }
    get title() { return this._title; }
    get text() { return this._text; }
    get drawingFile() { return this._drawingFile; }
    get drawingSize() { return this._drawingSize; }
    get drawing() { return this._drawing; }
    get idPlace() { return this._idPlace; }
    get labelPlace() { return this._labelPlace; }
    get creation() { return this._creation; }
    get lastUpdate() { return this._lastUpdate; }

    get hasDrawing() {
        return this._drawingFile !== undefined || this._drawing !== undefined;
    }

    get isEmpty() {
        return this._title === undefined && this._text === undefined && !this.hasDrawing;
    }

    /**
     * Bytes que ocupa la nota de cara a la cuota del usuario: el fichero del
     * dibujo (o el base64 heredado) más el texto guardado en el documento.
     */
    get size() {
        let size = 0;
        if (this._drawingSize !== undefined) {
            size += this._drawingSize;
        } else if (this._drawing !== undefined) {
            size += Buffer.byteLength(this._drawing, 'utf8');
        }
        if (this._title !== undefined) {
            size += Buffer.byteLength(this._title, 'utf8');
        }
        if (this._text !== undefined) {
            size += Buffer.byteLength(this._text, 'utf8');
        }
        return size;
    }

    toMap() {
        return {
            id: this._id,
            ...(this._title !== undefined && { title: this._title }),
            ...(this._text !== undefined && { text: this._text }),
            ...(this._drawingFile !== undefined && { drawingFile: this._drawingFile }),
            ...(this._drawingSize !== undefined && { drawingSize: this._drawingSize }),
            ...(this._drawing !== undefined && { drawing: this._drawing }),
            ...(this._idPlace !== undefined && { idPlace: this._idPlace }),
            ...(this._labelPlace !== undefined && { labelPlace: this._labelPlace }),
            creation: this._creation,
            lastUpdate: this._lastUpdate,
        };
    }
}

/**
 * Tamaño total en bytes que ocupan las notas de un usuario.
 * @param {Array} notes documentos de nota tal y como están en MongoDB
 */
function notesSize(notes) {
    if (!Array.isArray(notes)) {
        return 0;
    }
    return notes.reduce((total, note) => {
        try {
            return total + new Note(note).size;
        } catch {
            return total;
        }
    }, 0);
}

module.exports = { Note, notesSize };

File: client/lib/util/helpers/note.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:momoest/util/exceptions.dart';

/// Nota privada del estudiante. Puede tener título, texto y/o un dibujo, y
/// opcionalmente estar vinculada a un lugar.
///
/// El dibujo se guarda como fichero en el servidor: la nota solo conserva su
/// nombre ([drawingFile]) y su tamaño ([drawingSize]). [drawingLegacy] es el
/// PNG en base64 de las notas creadas antes de ese cambio, que se siguen
/// mostrando; [pendingDrawing] es un dibujo recién hecho todavía sin subir.
class Note {
  late String _id;
  String? title, text, idPlace, labelPlace;
  String? drawingFile, drawingLegacy;
  int? drawingSize;
  Uint8List? pendingDrawing;
  bool removeDrawing = false;
  late int creation, lastUpdate;
  Uint8List? _legacyBytes;

  Note.empty() {
    _id = '';
    creation = DateTime.now().millisecondsSinceEpoch;
    lastUpdate = creation;
  }

  Note(dynamic data) {
    if (data is Map) {
      if (data.containsKey('id') &&
          data['id'] is String &&
          data['id'].toString().trim().isNotEmpty) {
        _id = data['id'].toString().trim();
      } else {
        throw NoteException('id');
      }
      title = data['title'] is String && data['title'].trim().isNotEmpty
          ? data['title'].trim()
          : null;
      text = data['text'] is String && data['text'].trim().isNotEmpty
          ? data['text'].trim()
          : null;
      drawingFile =
          data['drawingFile'] is String && data['drawingFile'].trim().isNotEmpty
              ? data['drawingFile'].trim()
              : null;
      drawingSize = data['drawingSize'] is int ? data['drawingSize'] : null;
      drawingLegacy = data['drawing'] is String && data['drawing'].isNotEmpty
          ? data['drawing']
          : null;
      idPlace = data['idPlace'] is String && data['idPlace'].trim().isNotEmpty
          ? data['idPlace'].trim()
          : null;
      labelPlace =
          data['labelPlace'] is String && data['labelPlace'].trim().isNotEmpty
              ? data['labelPlace'].trim()
              : null;
      creation = data['creation'] is int
          ? data['creation']
          : DateTime.now().millisecondsSinceEpoch;
      lastUpdate = data['lastUpdate'] is int ? data['lastUpdate'] : creation;
      if (title == null && text == null && !hasDrawing) {
        throw NoteException('empty note');
      }
    } else {
      throw NoteException('Data is not a Map');
    }
  }

  String get id => _id;
  bool get hasId => _id.isNotEmpty;
  bool get hasPlace => idPlace != null;

  /// True si la nota tiene dibujo: uno nuevo sin subir, uno guardado como
  /// fichero o uno heredado en base64.
  bool get hasDrawing =>
      pendingDrawing != null ||
      (!removeDrawing && (drawingFile != null || drawingLegacy != null));

  /// True si el dibujo está guardado en el servidor y hay que descargarlo.
  bool get hasDrawingFile =>
      pendingDrawing == null && !removeDrawing && drawingFile != null;

  /// True si el dibujo es uno heredado, incrustado en el propio documento.
  bool get hasDrawingLegacy =>
      pendingDrawing == null && !removeDrawing && drawingLegacy != null;

  bool get isEmpty => title == null && text == null && !hasDrawing;

  /// Bytes del dibujo heredado, decodificados (con caché local)
  Uint8List? get legacyBytes {
    if (drawingLegacy == null) return null;
    _legacyBytes ??= base64Decode(drawingLegacy!);
    return _legacyBytes;
  }

  /// Registra un dibujo nuevo pendiente de subir; con null se quita el dibujo.
  void setDrawing(Uint8List? bytes) {
    pendingDrawing = bytes;
    removeDrawing = bytes == null;
  }

  /// Bytes que ocupa la nota, para poder mostrar el consumo antes de guardar.
  int get size {
    int size = pendingDrawing?.lengthInBytes ??
        (removeDrawing
            ? 0
            : drawingSize ?? (drawingLegacy != null ? drawingLegacy!.length : 0));
    if (title != null) size += utf8.encode(title!).length;
    if (text != null) size += utf8.encode(text!).length;
    return size;
  }

  /// Campos de la nota para el cuerpo de la petición. El campo `drawing` le
  /// dice al servidor si conserva o descarta el dibujo que ya tuviera guardado
  /// (se ignora cuando se envía un dibujo nuevo como fichero).
  Map<String, String> toFields() {
    return {
      if (title != null) 'title': title!,
      if (text != null) 'text': text!,
      if (idPlace != null) 'idPlace': idPlace!,
      if (labelPlace != null) 'labelPlace': labelPlace!,
      'drawing': removeDrawing ? 'remove' : 'keep',
    };
  }

  set id(String id) {
    if (id.trim().isNotEmpty) {
      _id = id.trim();
    }
  }
}
